### ISSUE TITLE
Warmup starts from lr=0

### ISSUE THREAD

**Issue by cfhammill** [pytorch/torchtune#2010]
Hi I'm using the `get_cosine_schedule_with_warmup` function from torchtune and according to my LR monitor the first epoch is training with LR=0. https://github.com/pytorch/torchtune/blob/main/torchtune/training/lr_schedulers.py#L43

Assuming that step indexes from zero in the code above it should indeed return an lr multiplicative factor of zero. If this is the case and I understand correctly that means there is a epoch where no learning happens at the start of a training run. 

I'm curious if this is correct and intentional.

**Comment by felipemello1** [comment]
hey @cfhammill, good question! We update the lr on every step. One step is one forward+backward pass o the model (assuming grad_acc_steps=1). You can see where we call it: https://github.com/pytorch/torchtune/blob/bd89bb88b6092de74e79356ddb193b9e4b614d46/recipes/lora_finetune_single_device.py#L720

But you are right that the first step has lr=0 and that doesn't sound right. It should start from a min_lr.

**Comment by felipemello1** [comment]
Let me know if you would like to submit a PR to address the issue. Otherwise i can create a ticket for us to follow up. :)

**Comment by cfhammill** [comment]
Thanks @felipemello1, if the lr scheduler is intended to be stepped every optimizer step, losing the first batch or accumulated batch is probably not the end of the world. That being said, I think some training frameworks, including pytorch-lightning, step the scheduler every epoch, not every optimizer step. The torch docs for LambdaLR show stepping the scheduler every epoch as their example:

https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html#:~:text=the%20learning%20rate.-,Example,-%3E%3E%3E%20%23%20Assuming%20optimizer

so probably worth fixing for people like me who are using lightning.

Happy to work up a PR. I can think of two simple fixes, one would be to add one to the step index and the total warmup steps in the calculation here

https://github.com/pytorch/torchtune/blob/main/torchtune/training/lr_schedulers.py#L46

the other would be to add an additional argument, e.g. `min_lr` defaulting to `1 / num_warmup_steps` and the returned lr factor can be wrapped in `max(min_lr, ...)`. This could optionally also be added to post warmup phase replacing zero in

https://github.com/pytorch/torchtune/blob/main/torchtune/training/lr_schedulers.py#L56

I think for simplicity and the sake of not adding an additional argument, the former approach is probably sufficient

**Comment by felipemello1** [comment]
hey @cfhammill , i believe its ok to call it on every training step. It is actually the first example in lightining: https://lightning.ai/docs/pytorch/stable/common/optimization.html#learning-rate-scheduling 

<img width="497" alt="image" src="https://github.com/user-attachments/assets/3ef604f1-df82-41d3-9026-e4f795d3947a">

Specially for training LLMs, i think that this is the way to go, since most of the people train it for just a few epochs. I rarely see >10 epochs on a dataset. If we did a step for every epoch, then the LR change wouldn't be smooth (think of epoch=3 being a triangle). With that being said, It should be fine to give the option to the user to update on the step or the epoch, but I don't think that it should make a big difference. Let me know if you have any thoughts here.

> losing the first batch or accumulated batch is probably not the end of the world

agreed

> I think for simplicity and the sake of not adding an additional argument, the former approach is probably sufficient

This should be ok. But if you have the time, ideally we should:
1. Plot the lr graph generated by the scheduler
2. Check if there are many points in the long tail that have LR that are just too low to matter, e.g. <1e-6. If thats the case, then i think that the right thing to do is to add min_lr.
3. In case we go with 2, then we could add the min_lr with a reasonable default, ie 1e-7.

> Happy to work up a PR

That's awesome! I will leave it up if you want to just add +1 to the index, or if you want to investigate adding min_lr. Either way, please run it with any dummy model for a few steps and paste the LR curve, so can make sure that nothing broke.

More info on contributing here, but the TLDR:
https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md 

1) Fork torchtune: https://github.com/pytorch/torchtune
git clone https://github.com/<YOUR_GITHUB_USER>/torchtune.git

2) install dependencies

cd torchtune
conda create -n torchtune python=3.11
conda activate torchtune
pip install --pre --upgrade torch torchvision torchao --index-url https://download.pytorch.org/whl/nightly/cu124
pip install -e ".[dev]"
pre-commit install

3) create the PR

**Comment by cfhammill** [comment]
Thanks @felipemello1, the default behaviour in lightning is to handle your optimization for you, and the default config is to step every epoch
https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.configure_optimizers:~:text=The%20lr_scheduler_config%20is%20a%20dictionary%20which%20contains%20the%20scheduler%20and%20its%20associated%20configuration.%20The%20default%20configuration%20is%20shown%20below.. you can change this to every step with an entry in the config dict or to use manual mode as you show above. So yes it's definitely possible to update every step.

Good point regarding incrementing every step making much more sense for llm training/tuning.

I like the idea of using a small constant as the min lr, and having it as an option. I'll take that approach. I don't think I need to run a model to plot the LR, making the scheduler and stepping it for a large number of steps should be sufficient right?

 Do you have an estimate of how many steps are typical for modern text datasets?

Thanks for the contributing instructions. Will work this up soon.

**Comment by felipemello1** [comment]
> Do you have an estimate of how many steps are typical for modern text datasets? 

I think that its reasonable to imagine a dataset with 100k examples and a batch_size=8, so 100k/8 steps. But it shouldn't matter too much, you can use a size=1k.

>  I don't think I need to run a model to plot the LR, making the scheduler and stepping it for a large number of steps should be sufficient right?

for developing, yes, but as a sanity check, after you are done, running the model should be the way to go. You can do something as simple as the code below with qwen 0.5B or llama 3.2 1b. If you don't have the budget, thats fine. I can run it after you submit the PR.

```
tune run <...> \
metric_logger=torchtune.training.metric_logging.WandBLogger \
max_steps_per_epoch=10 \
epochs=2 \
lr_scheduler._component_=torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup
lr_scheduler.num_warmup_steps=5
```

> I like the idea of using a small constant as the min lr, and having it as an option

Awesome! I would just avoid be careful with having flat lines of LR because of using max(min_lr, some_other_lr). Ideally, the scheduler should start from min_lr. Eg:

```
min_lr=1
original: -2.0, -1.0, 0, 1.0, 2.0, 3.0, 4.0
bad:         1.0, 1.0, 1.0, 1.0, 2.0, 3.0, 4.0
good:       1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0
```

pytorch already offers some APIs, so maybe there is something there that we can use and keep the code simple.

**Comment by felipemello1** [comment]
@cfhammill I was just checking it, and actually the first step doesn't have lr=0. I believe LambdaLR uses first step = 1. Also, the first LR is reasonable, 3e-6 for 100 warmup steps.

My guess is that its **not** worth going down this rabbit hole:
<img width="305" alt="image" src="https://github.com/user-attachments/assets/1f82a8dd-5de4-422d-a1f7-170d272b1991">

Please feel free to investigate it quickly and drop if it thats the case.

**Comment by cfhammill** [comment]
I think we'd probably be stuck in the flat LR regime if we choose a constant min_lr out in the tails. For example if we did 10000 steps of post warmup training, the last 6 steps would be below 1e-6

https://github.com/pytorch/torchtune/blob/main/torchtune/training/lr_schedulers.py#L53-55

inlining a half cycle and progress as described above we , so step 9994-10000 would all be below threshold

```
import math
cosine_lr_multiple = 0.5 * (
            1.0 + math.cos(math.pi * 0.5 * 2.0 * 9994/10000)
)
```
~= 8.88e-07

and also note this just the LR factor, if base LR is < 1 the actual LR is proportionally smaller. I have no idea how small an effective LR is worth training at, but it at least seems flat regions in the tail are unavoidable.

Note that for warmup the situation is fine, you get one step of `min_lr` and the your next step is 1/warmup_steps which will likely be much much larger. If `min_lr` is so small as to be effectively zero, maybe this change isn't worthwhile (although at least it is then configurable).

Regarding you more recent message, I don't think that's true:


```
from torch.optim.lr_scheduler import LambdaLR
from torch.optim import SGD

schedule = LambdaLR(SGD(params = [torch.tensor(1)]), lr_lambda = lambda e: e)

print(schedule.get_last_lr())
schedule.step()
print(schedule.get_last_lr())
```

prints 0, 0.001 for me (the default base_lr for SGD is 1e-3)

**Comment by felipemello1** [comment]
> I have no idea how small an effective LR is worth training at, but it at least seems flat regions in the tail are unavoidable.

I see. Yeah, in this case flat seems fine to me. In my experience, anything below 1e-6 won't significantly move the loss.

**Comment by joecummings** [comment]
Closing as complete!
